File: health_prompt/HealthDashboard/Models/HDHealthDataModel.m
//
//  HDHealthDataModel.m
//  HealthDashboard - 健康数据模型
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDHealthDataModel.h"

@implementation HDMoodRecord
- (NSString *)emojiString {
    NSArray *emojis = @[@"😞",@"😕",@"😐",@"😊",@"😄"];
    NSInteger idx = MAX(0, MIN(4, self.moodLevel - 1));
    return emojis[idx];
}
@end

@implementation HDHealthDataModel

+ (instancetype)shared {
    static HDHealthDataModel *_inst;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{ _inst = [self new]; });
    return _inst;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        // 初始化模拟数据
        _todaySteps  = 6842;
        _stepsGoal   = 10000;
        _waterML     = 600;
        _waterGoalML = 2000;
        _sleepHours  = @[@6.5, @7.2, @5.8, @8.0, @7.5, @6.0, @7.8];
        _isDarkMode  = NO;

        // 模拟心情记录
        NSMutableArray *moods = [NSMutableArray array];
        NSInteger levels[] = {3,4,2,5,3,4,4};
        for (int i = 6; i >= 0; i--) {
            HDMoodRecord *r = [HDMoodRecord new];
            r.moodLevel  = levels[6-i];
            r.timestamp  = [NSDate dateWithTimeIntervalSinceNow:-i*86400];
            [moods addObject:r];
        }
        _moodRecords = moods;
    }
    return self;
}

- (CGFloat)stepsProgress { return MIN(1.0, (CGFloat)_todaySteps / _stepsGoal); }
- (CGFloat)waterProgress { return MIN(1.0, _waterML / _waterGoalML); }
- (HDMoodRecord *)latestMood { return _moodRecords.lastObject; }

- (void)addWater:(CGFloat)ml {
    _waterML = MIN(_waterGoalML, _waterML + ml);
    [[NSNotificationCenter defaultCenter] postNotificationName:@"HDDataDidChange" object:nil];
}

- (void)addSteps:(NSInteger)steps {
    _todaySteps += steps;
    [[NSNotificationCenter defaultCenter] postNotificationName:@"HDDataDidChange" object:nil];
}

- (void)addMood:(NSInteger)level {
    HDMoodRecord *r = [HDMoodRecord new];
    r.moodLevel  = level;
    r.timestamp  = [NSDate date];
    NSMutableArray *arr = [_moodRecords mutableCopy];
    [arr addObject:r];
    _moodRecords = arr;
    [[NSNotificationCenter defaultCenter] postNotificationName:@"HDDataDidChange" object:nil];
}

- (CGFloat)caloryForSteps:(NSInteger)steps {
    // 平均每步消耗约0.04千卡
    return steps * 0.04;
}

@end

File: health_advanced_swift/HealthDashboard/Models/HDHealthDataModel.swift
//
//  HDHealthDataModel.swift
//  HealthDashboard - 健康数据模型
//
//  Created by AI Assistant on 2026/3/22.
//

import UIKit
import Combine

// MARK: - HDMoodRecord

/// 心情记录
class HDMoodRecord {
    var moodLevel: Int
    var timestamp: Date
    var emojiString: String {
        let emojis = ["😞", "😕", "😐", "😊", "😄"]
        return emojis[max(0, min(4, moodLevel - 1))]
    }
    init(moodLevel: Int = 3, timestamp: Date = Date()) {
        self.moodLevel = moodLevel
        self.timestamp = timestamp
    }
}

// MARK: - HDExerciseRecord

/// 运动记录
class HDExerciseRecord {
    var type: Int = 0
    var durationSeconds: Int = 0
    var distanceKM: CGFloat = 0
    var caloriesBurned: Int = 0
    var timestamp: Date = Date()
    var typeString: String { type == 0 ? "目标跑" : "自由跑" }
}

// MARK: - HDHealthDataModel

/// 全局健康数据单例（ObservableObject，@Published 驱动 UI）
@MainActor
final class HDHealthDataModel: ObservableObject {

    // MARK: - Singleton

    static let shared = HDHealthDataModel()
    private init() {}

    // MARK: - 步数

    @Published var todaySteps: Int = 6842
    @Published var stepsGoal: Int = 10000
    var stepsProgress: CGFloat { min(1.0, CGFloat(todaySteps) / CGFloat(stepsGoal)) }

    // MARK: - 喝水

    @Published var waterML: CGFloat = 600
    @Published var waterGoalML: CGFloat = 2000
    var waterProgress: CGFloat { min(1.0, waterML / waterGoalML) }

    // MARK: - 睡眠

    @Published var sleepHours: [Double] = [6.5, 7.2, 5.8, 8.0, 7.5, 6.0, 7.8]

    // MARK: - 心情

    @Published var moodRecords: [HDMoodRecord] = {
        let levels = [3, 4, 2, 5, 3, 4, 4]
        return (0..<7).map { i in
            HDMoodRecord(moodLevel: levels[i],
                         timestamp: Date(timeIntervalSinceNow: -Double(6 - i) * 86400))
        }
    }()
    var latestMood: HDMoodRecord? { moodRecords.last }

    // MARK: - 运动

    @Published var targetRunDistanceKM: Int = 5
    @Published var targetRunMinutes: Int = 30
    @Published var exerciseHistory: [HDExerciseRecord] = []

    // MARK: - 主题

    @Published var isDarkMode: Bool = false

    // MARK: - 操作方法

    /// 添加饮水记录
    func addWater(_ ml: CGFloat) {
        waterML = min(waterGoalML, waterML + ml)
    }

    /// 添加步数
    func addSteps(_ steps: Int) {
        todaySteps += steps
    }

    /// 添加心情记录
    func addMood(_ level: Int) {
        moodRecords.append(HDMoodRecord(moodLevel: level))
    }

    /// 保存运动记录
    func saveExerciseRecord(_ record: HDExerciseRecord) {
        exerciseHistory.append(record)
    }

    /// 步数换算卡路里（约0.04千卡/步）
    func calory(forSteps steps: Int) -> CGFloat {
        CGFloat(steps) * 0.04
    }
}
